docs: Oryx components (#2187)
spryker/spryker-docs: docs/scos/dev/front-end-development/202307.0/oryx/building-components/oryx-implementing-components.md

Styling components in the shadow DOM is a big topic we recommend studying separately. However, there are a few things to know when it comes to Oryx and styling components:

- Design system components are provided in the UI package. Components like `<oryx-button>` or `<oryx-link>` are used to ensure a common visual language. They can be customized.
- Font styles rules like `font-face` or `font-size`, unlike other CSS rules, cascade to web components, no matter how deep they are nested. Standard font rules are provided in the `<oryx-app>` component.
- [Typography](/docs/scos/dev/front-end-development/{{page.version}}/oryx/building-applications/styling/oryx-typography.html) design tokens are used to ensure consistent styling.
- Custom properties, also known as CSS variables, cascade into web components, which is why the application theme is based on CSS variables. See [Design tokens](/docs/scos/dev/front-end-development/{{page.version}}/oryx/building-applications/styling/oryx-design-tokens.html) for more information.

> - Oryx provides an [icon system](/docs/scos/dev/front-end-development/{{page.version}}/oryx/building-applications/styling/oryx-icon-system.html) that you can leverage in your components.

Reviewer: Rephrase this to make it relevant to styling the component

Author: rephrased a bit, but I was quit ok with the text already. We might have a disconnect here.

Localizations are resolved asynchronously, and require the UI to be rerendered whenever they're loaded or reloaded. `ContentMixin`, which you've used earlier to integrate the component options, provides access to the `i18n` directive. The `i18n` directive is available as a class method. The following example shows how to use it:

```ts
protected render(): TemplateResult | void {
  return html`${this.i18n('cart.add-to-cart')}`;
}
```

If you do not use `ContentMixin`, you can use `I18nMixin` instead. If you choose to not use mixins, you can integrate the `i18n` directive directly.

## 6. Using services inside the component

You've seen how `ProductMixin` resolves the product data and hides the integration with the `ProductService`. It is also common to use services directly in components. Oryx _injects_ services using [dependency injection (DI)](/docs/scos/dev/front-end-development/{{page.version}}/oryx/architecture/dependency-injection/dependency-injection.html). DI provides decoupling of components and shared business logic. This is a common design pattern that separates concerns and lets you customize services without touching the components or other depending services.

> The Oryx DI container is used to register and resolve services by a token. You can read more about resolving services in [the documentation](/docs/scos/dev/front-end-development/{{page.version}}/oryx/architecture/dependency-injection/dependency-injection-using-services.html). In the following example you see how the pricing service is resolved.

Reviewer: ```suggestion
The Oryx DI container is used to register and resolve services using a token. You can read more about resolving services in [the documentation](/docs/scos/dev/front-end-development/{{page.version}}/oryx/architecture/dependency-injection/dependency-injection-using-services.html). In the following example you see how the pricing service is resolved.
```

When developing a component, you need to configure the hydration trigger using the `@hydrate` decorator that can take an event or context. The following example shows how to set up the component to be hydrated when the context is changed:

```ts
import { resolve } from "@spryker-oryx/di";
import { ProductMixin } from "@spryker-oryx/product";
import { hydrate } from "@spryker-oryx/utilities";

@hydrate({ context: ProductContext.SKU })
export class ProductIdComponent extends ProductMixin(LitElement) {
  // ...
}
```

Alternatively, you can configure hydration to be triggered by a specific event:

```ts
@hydrate({ event: ["mouseover", "focus"] })
export class ProductIdComponent extends ProductMixin(LitElement) {
  // ...
}
```

> ## Registering the component definition

Author: No, the previous steps are steps of the first section. I've made them level 3 sections, hope that works for you.

---
title: "Oryx: Implementing components"
description: Learn how to create an Oryx component
last_updated: Sept 20, 2023
template: concept-topic-template
---

Oryx components are web components built with [Lit](https://lit.dev). Lit is a lightweight open-source framework from Google that's used to build highly efficient web components. Web components can be created with any framework or even with vanilla HTML, CSS, and JavaScript. You can use any other framework instead of Lit. However, some Oryx utilities, like [signals](docs/scos/dev/front-end-development/{{oage.version}}/oryx/architecture/reactivity/signals.html) and component mixins, are available only with Lit.

## Implementing a component

This document describes how to implement a component using Lit. Lit's standard concepts are not covered. To learn more about them, see [Components overview](https://lit.dev/docs/components/overview/).

We use the product _ID_ component as an example. It is a simple component that shows the basic concepts. The component already exists in the Oryx product package.

> Oryx creates a folder per component, like `src/product/id`, and separates some of the component logic in separate files. However, you can create a component as a single file. To allow for lazy loading of the component, you still need to separate out its definition.

Reviewer: ```suggestion
In Oryx, components are organized in folders, like `src/product/id`, with some component logic located in separate files. However, you can create a component as a single file. To allow for lazy loading of the component, you still need to separate out its definition.
```
